server/src/main/java/org/apache/cloudstack/diagnostics/RemoteDiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.log4j.Logger;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class RemoteDiagnosticsServiceImpl extends ManagerBase implements PluggableService, RemoteDiagnosticsService, Configurable {
+    private static final Logger s_logger = Logger.getLogger(RemoteDiagnosticsServiceImpl.class);
+
+    @Inject
+    private RouterControlHelper routerControlHelper;
+    @Inject
+    private AgentManager agentManager;
+    @Inject
+    private VMInstanceDao vmInstanceDao;
+    @Inject
+    private ConfigurationDao configurationDao;
+
+
+
+    @Override
+    public RemoteDiagnosticsResponse executeDiagnosticsToolInSystemVm(final RemoteDiagnosticsCmd cmd) throws AgentUnavailableException,
+            InvalidParameterValueException {
+        final Long systemVmId = cmd.getId();
+        final VMInstanceVO systemVm = vmInstanceDao.findByIdTypes(systemVmId, VirtualMachine.Type.ConsoleProxy,
+                VirtualMachine.Type.DomainRouter, VirtualMachine.Type.SecondaryStorageVm);
+        if (systemVm == null){
+            s_logger.error("Invalid system VM id provided " + systemVmId);
+            throw new InvalidParameterValueException("Unable to find a system virtual machine with id " + systemVmId);
+        }
+
+        final Long hostId = systemVm.getHostId();
+        if (hostId == null){
+            s_logger.warn("Unable to find host for virtual machine instance " + systemVm.getInstanceName());
+            throw new CloudRuntimeException("Unable to find host for virtual machine instance " + systemVm.getInstanceName());
+        }
+
+        final String cmdType = cmd.getType();
+        final String cmdAddress = cmd.getAddress();
+        final String optionalArgunments = cmd.getOptionalArguments();
+
+        String remoteCommand = setupRemoteCommand(cmdType, cmdAddress, optionalArgunments);
+
+        final ExecuteDiagnosticsCommand command = new ExecuteDiagnosticsCommand(remoteCommand);
+        command.setAccessDetail(NetworkElementCommand.ROUTER_IP, routerControlHelper.getRouterControlIp(systemVm.getId()));
+        command.setAccessDetail(NetworkElementCommand.ROUTER_NAME, systemVm.getInstanceName());
+
+        Answer origAnswer;
+        try{
+            origAnswer = agentManager.send(hostId, command);
+        } catch (final OperationTimedoutException e) {
+            s_logger.warn("Timed Out", e);
+            throw new AgentUnavailableException("Unable to send commands to virtual machine ", hostId, e);
+        }
+
+        ExecuteDiagnosticsAnswer answer = null;
+        if (origAnswer instanceof ExecuteDiagnosticsAnswer) {
+            answer = (ExecuteDiagnosticsAnswer) origAnswer;
+        }
+
+        return createRemoteDiagnosisResponse(answer);
+    }
+
+    protected RemoteDiagnosticsResponse createRemoteDiagnosisResponse(ExecuteDiagnosticsAnswer answer){
+        RemoteDiagnosticsResponse response = new RemoteDiagnosticsResponse();
```

Comment:
Can you move this to the Cmd class. Usually it's a good idea for the cmd class to create response, but this is okay as well.

Reply:
Refactored to Cmd class